### PR TITLE
Now allow specification of the number of cores for benchmark plotting jobs

### LINKED ISSUE #179: [FEATURE REQUEST] Add switch to quickly disable parallelization in gcpy
Currently, to disable parallelization, one has to manually go into the code and change `Parallel(n_jobs=-1)` to `Parallel(n_jobs=1)` where applicable. To simplify debugging, we should add a switch to the config file to easily disable parallelization throughout.

### PR DESCRIPTION
**Pull request by yantosca**
### Description
This PR does the following:

1. Adds the `options.n_cores` YAML tag to benchmark YAML configuration files, in the `benchmark/config` and `benchmark/cloud` folders, e.g.:
```yaml
#
# options: Customizes the benchmark plot output.
#
options:
  #
  # bmk_type: Specifies the type of benchmark.
  #
  bmk_type: FullChemBenchmark
  #
  # comparisons: Specifies the comparisons to perform.
  #
  comparisons:
  ... etc not shown ...
  #
  # outputs: Specifies the plots and tables to generate.
  #
  outputs:
  ... etc not shown ...
  #
  # n_cores: Specify the number of cores to use:
  # -1: Use as many cores as possible
  #  N: Use N cores
  #  1: Disable parallelization (use a single core)
  #
  n_cores: -1
```

2. Reads `options.n_cores` into the `config["options"]["n_cores"]` dict entry.

3. Passes `n_job=config["options"]["n_cores"]` to the relevant benchmark plotting routines in the following modules:
- `benchmark/run_benchmark.py`
- `benchmark/modules/run_1yr_fullchem_benchmark,py`
- `benchmark/modules/run_1yr_tt_benchmark.py`

### Related issues:

- Closes #179

### CODE REVIEW

**Comment by yantosca** [comment]
@laestrada and I have discovered that the joblib `Parallel` function can potentially mask errors in other areas of code.  For this reason, I propose to rewrite the code so that we use an if block to execute without parallelization if n_job=1, or to call the joblib `Parallel` function otherwise.  Stay tuned.

**Comment by yantosca** [comment]
We have now added `if` blocks around the calls to `joblib.Parallel()` so that separate, non-parallelized code will execute when `n_job==1`, such as shown in the example below:

```python
        # ==================================================================
        # GCC vs GCC operations budgets tables
        # ==================================================================
        if config["options"]["outputs"]["ops_budget_table"]:
            print("\n%%% Creating GCC vs. GCC operations budget tables %%%")

            def gcc_vs_gcc_ops_budg(mon):
                """
                Create budget table for each benchmark month m in parallel
                """

                # Filepaths
                refpath = get_filepath(
                    gcc_vs_gcc_refdir,
                    "Budget",
                    bmk_mons_ref[mon]
                )
                devpath = get_filepath(
                    gcc_vs_gcc_devdir,
                    "Budget",
                    bmk_mons_dev[mon]
                )

                # Create tables
                bmk.make_benchmark_operations_budget(
                    config["data"]["ref"]["gcc"]["version"],
                    refpath,
                    config["data"]["dev"]["gcc"]["version"],
                    devpath,
                    sec_per_month_ref[mon],
                    sec_per_month_dev[mon],
                    benchmark_type=bmk_type,
                    label=f"at 01{bmk_mon_yr_strs_dev[mon]}",
                    dst=gcc_vs_gcc_tablesdir,
                )

            # Create tables in parallel
            # Turn off parallelization if n_jobs==1
            if n_jobs != 1:
                results = Parallel(n_jobs=config["options"]["n_cores"])(
                    delayed(gcc_vs_gcc_ops_budg)(mon) \
                    for mon in range(bmk_n_months)
                )
            else:
                for mon in range(bmk_n_months):
                    results = gcc_vs_gcc_ops_budg(mon)
```
@laestrada and I have found that `joblib.Parallel()` often obfuscates error messages from other Python packages.  For the sake of easy debugging, we no longer rely on `joblib.Parallel(n_jobs=1)` to disable parallelization.

**Comment by yantosca** [comment]
This PR should be merged before PR #250.

File: benchmark/cloud/template.1hr_benchmark.yml [hunk 1]
```diff
@@ -110,3 +120,10 @@ options:
     plot_options:
       by_spc_cat: True
       by_hco_cat: True
+  #
+  # n_cores: Specify the number of cores to use:
+  # -1: Use all available cores
+  # -N: Use N cores
```
Comment (msulprizio) [inline benchmark/cloud/template.1hr_benchmark.yml:126]: Should this be `N` not `-N`? Same comment applies to rest of yml files modified here.

**Review by msulprizio — CHANGES_REQUESTED** [review]

File: benchmark/cloud/template.1hr_benchmark.yml (hunk 1 above)
Comment (yantosca) [reply benchmark/cloud/template.1hr_benchmark.yml:126]: That's what I thought but the manual for joblib.Parallel says:

>**Parameters**
>n_jobs: int, default: None
The maximum number of concurrently running jobs, such as the number of Python worker processes when backend=”multiprocessing” or the size of the thread-pool when backend=”threading”. If -1 all CPUs are used. If 1 is given, no parallel computing code is used at all, and the behavior amounts to a simple python for loop. This mode is not compatible with timeout. For n_jobs below -1, (n_cpus + 1 + n_jobs) are used. Thus for n_jobs = -2, all CPUs but one are used. None is a marker for ‘unset’ that will be interpreted as n_jobs=1 unless the call is performed under a [parallel_config()](https://joblib.readthedocs.io/en/stable/generated/joblib.parallel_config.html#joblib.parallel_config) context manager that sets another value for n_jobs.

In practice you would use -1 (all cores) or 1 (no cores).  I can update the YAML comments files accordingly

Comment (yantosca) [reply benchmark/cloud/template.1hr_benchmark.yml]: In commit be766a2, all benchmark YAML files have been updated to use these comments, which are more accurate:
```yaml
  #
  # n_cores: Specify the number of cores to use.
  # -1: Use $OMP_NUM_THREADS         cores
  # -2: Use $OMP_NUM_THREADS - 1     cores
  # -N: Use $OMP_NUM_THREADS - (N-1) cores
  #  1: Disable parallelization (use a single core)
  #
  n_cores: -1
```

**Pull request merged**